br/pkg/lightning/backend/external/onefile_writer.go

```diff
@@ -0,0 +1,235 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package external
+
+import (
+	"context"
+	"encoding/binary"
+	"encoding/hex"
+	"path/filepath"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/br/pkg/lightning/common"
+	"github.com/pingcap/tidb/br/pkg/storage"
+	tidbkv "github.com/pingcap/tidb/pkg/kv"
+	"github.com/pingcap/tidb/pkg/util/logutil"
+	"github.com/pingcap/tidb/pkg/util/size"
+	"go.uber.org/zap"
+)
+
+// OneFileWriter is used to write data into external storage with one file.
+type OneFileWriter struct {
+	store          storage.ExternalStorage
+	writerID       string
+	filenamePrefix string
+	keyAdapter     common.KeyAdapter
+
+	kvStore *KeyValueStore
+	rc      *rangePropertiesCollector
+
+	kvBuffer    *preAllocKVBuf
+	kvLocations []kvLocation
+	kvSize      int64
+
+	onClose OnCloseFunc
+	closed  bool
+
+	// Statistic information per batch.
+	batchSize uint64
+
+	multiFileStat MultipleFilesStat
+	// Statistic information per writer.
+	minKey    tidbkv.Key
+	maxKey    tidbkv.Key
+	totalSize uint64
+
+	dataFile   string
+	statFile   string
+	dataWriter storage.ExternalFileWriter
+	statWriter storage.ExternalFileWriter
+	logger     *zap.Logger
+}
+
+// initWriter inits the underlying dataFile/statFile path, dataWriter/statWriter for OneFileWriter.
+func (w *OneFileWriter) initWriter(ctx context.Context, partSize int64) (
+	err error,
+) {
+	w.dataFile = filepath.Join(w.filenamePrefix, "one-file")
+	w.dataWriter, err = w.store.Create(ctx, w.dataFile, &storage.WriterOption{Concurrency: 20, PartSize: partSize})
+	if err != nil {
+		return err
+	}
+	w.statFile = filepath.Join(w.filenamePrefix+statSuffix, "one-file")
+	w.statWriter, err = w.store.Create(ctx, w.statFile, &storage.WriterOption{Concurrency: 20, PartSize: int64(5 * size.MB)})
+	if err != nil {
+		_ = w.dataWriter.Close(ctx)
+		return err
+	}
+	w.logger.Info("one file writer", zap.String("data-file", w.dataFile), zap.String("stat-file", w.statFile))
+	return nil
+}
+
+// Init inits the OneFileWriter and its underlying KeyValueStore.
+func (w *OneFileWriter) Init(ctx context.Context, partSize int64) (err error) {
+	w.logger = logutil.Logger(ctx)
+	err = w.initWriter(ctx, partSize)
+	if err != nil {
+		return err
+	}
+	w.kvStore, err = NewKeyValueStore(ctx, w.dataWriter, w.rc)
+	return err
+}
+
+// WriteRow implements ingest.Writer.
+func (w *OneFileWriter) WriteRow(ctx context.Context, idxKey, idxVal []byte) error {
+	keyAdapter := w.keyAdapter
+	var rowID []byte
+	encodedKeyLen := keyAdapter.EncodedLen(idxKey, rowID)
```

Comment:
as data is already sorted, seems we don't need to keep it in memory, we can write to `kvStore` directly.

Reply:
Great, I will fix it

Reply:
> collect stat logic

this part is in kvStore, we can still reuse it

Reply:
Yep, I mean maxKey and minKey